src/Toolkit/I18n.php

```diff
@@ -98,107 +89,158 @@ public static function formatNumber(int|float $number, string $locale = null): s
-
-		if (is_array($key) === true) {
-			// try to use actual locale
-			if ($result = $key[$locale] ?? null) {
-				return $result;
-			}
-			// try to use language code, e.g. `es` when locale is `es_ES`
-			if ($result = $key[Str::before($locale, '_')] ?? null) {
+		// create shorter locale code, e.g. `es` for `es_ES` locale
+		$shortLocale = Str::before($locale, '_');
+
+		// There are two main use cases that we will treat separately:
+		// (1) with a string representing an i18n key to be looked up
+		// (2) an array with entries per locale
+		//
+		// Both with various ways of handling fallbacks, provided
+		// explicitly via the parameter and/or from global defaults.
+
+		// (1) string $key: look up i18n string from global translations
+		if (is_string($key) === true) {
+			// look up locale in global translations list,
+			if ($result = static::translation($locale)[$key] ?? null) {
 				return $result;
 			}
-			// use global wildcard as i18n key
-			if (isset($key['*']) === true) {
-				return static::translate($key['*'], $key['*']);
+
+			// prefer any direct provided $fallback
+			// over further fallback alternatives
+			if ($fallback !== null) {
+				if (is_array($fallback) === true) {
+					return static::translate($fallback, null, $locale);
+				}
+
+				return $fallback;
 			}
-			// use fallback
-			if (is_array($fallback) === true) {
-				return
-					$fallback[$locale] ??
-					$fallback['en'] ??
-					reset($fallback);
+
+			// last resort: try using the fallback locales
+			foreach (static::fallbacks() as $fallback) {
+				// skip locale if we have already tried to save performance
+				if ($locale === $fallback) {
+					continue;
+				}
+
+				if ($result = static::translation($fallback)[$key] ?? null) {
+					return $result;
+				}
 			}
 
-			return $fallback;
+			return null;
 		}
 
-		// $key is a string
-		if ($result = static::translation($locale)[$key] ?? null) {
+		// --------
+		// (2) array|null $key with entries per locale
+
+		// try entry for long and short locale
+		if ($result = $key[$locale] ?? null) {
 			return $result;
 		}
-
-		if ($fallback !== null) {
-			return $fallback;
+		if ($result = $key[$shortLocale] ?? null) {
+			return $result;
 		}
 
-		foreach (static::fallbacks() as $fallback) {
-			// skip locales we have already tried
-			if ($locale === $fallback) {
-				continue;
+		// if the array as a global wildcard entry,
+		// use this one as i18n key and try to resolve
+		// this via part (1) of this method
+		if ($wildcard = $key['*'] ?? null) {
+			if ($result = static::translate($wildcard, $wildcard, $locale)) {
+				return $result;
 			}
+		}
+
+		// if the $fallback parameter is an array, we can assume
+		// that it's also an array with entries per locale:
+		// check with long and short locale if we find a matching entry
+		if ($result = $fallback[$locale] ?? null) {
+			return $result;
+		}
+		if ($result = $fallback[$shortLocale] ?? null) {
+			return $result;
+		}
```

Comment:
```suggestion
		if ($result = $fallback[$locale] ?? $fallback[$shortLocale] ?? null) {
			return $result;
		}
```

Reply:
This one I don't wanna do. I kept every possibility now as its own if Statement because that way test coverage really can show us if we indeed have each and every outcome covered by tests.

I really love chaining things with ?? But for that reason I removed it a lot in this method.